File: android/app/src/main/java/com/acquiring/android/HarvestLedger.kt
package com.acquiring.android

import android.database.sqlite.SQLiteDatabase
import androidx.room.withTransaction
import java.io.File

/**
 * Keeps manually harvested songs alive across a catalog install.
 *
 * DatabaseDownloader replaces `acquiring-db` as a whole file, so a harvested row
 * written only into the catalog is destroyed by every "Download Full Library".
 * The ledger in [UserDataDatabase] is what survives that swap, and [replay]
 * rebuilds from it whatever the new catalog does not carry itself.
 *
 * Ordering is not negotiable: [adopt] must run before the swap, because it reads
 * the outgoing catalog, and [replay] after it.
 */
object HarvestLedger {

    private const val LEGACY_ADOPTION_KEY = "legacyHarvestsAdopted"

    /**
     * A diff this large is not a set of manual harvests — it is a catalog that
     * was re-keyed or rebuilt. Adopting it would pin thousands of stale rows
     * into every future catalog, so the sweep declines instead.
     */
    internal const val LEGACY_ADOPTION_LIMIT = 500

    /** Records a harvest as it is written into the catalog. */
    suspend fun record(
        userDb: UserDataDatabase,
        song: Song,
        alphaGroup: String,
        modes: Iterable<String>
    ) {
        userDb.harvestLedgerDao().record(
            HarvestedSong(
                slug = song.slug,
                artist = song.artist,
                title = song.title,
                url = song.url,
                status = song.status,
                dataBlob = song.dataBlob ?: return,
                alphaGroup = alphaGroup,
                modes = HarvestedSong.encodeModes(modes),
                harvestedAt = System.currentTimeMillis()
            )
        )
    }

    /**
     * One-time amnesty for songs harvested before this ledger shipped: the
     * outgoing catalog is the only record of them, and the swap deletes it.
     *
     * Running it on every install would permanently resurrect songs the catalog
     * later prunes on purpose, so a flag beside the rows closes it afterwards.
     * "Looks harvested" is the absence of a complexity rating — every shipped
     * catalog row has one, and a harvest never computes one.
     */
    suspend fun adopt(currentDb: AppDatabase, stagedDbFile: File, userDb: UserDataDatabase) {
        val dao = userDb.harvestLedgerDao()
        if (dao.hasFlag(LEGACY_ADOPTION_KEY)) return

        // Slugs first, payloads only for the survivors: the candidate set is
        // bounded but its chord blobs are not, and most of them are about to be
        // discarded as "the new catalog has this one too".
        val candidateSlugs = currentDb.songDao().getUnratedSlugsWithChords(LEGACY_ADOPTION_LIMIT + 1)
        if (candidateSlugs.size > LEGACY_ADOPTION_LIMIT) return
        val stagedSlugs = slugsPresent(stagedDbFile, candidateSlugs)
        val adoptable = candidateSlugs.filter { it !in stagedSlugs }

        val now = System.currentTimeMillis()
        val entries = adoptable.mapNotNull { slug ->
            val song = currentDb.songDao().getSongBySlug(slug) ?: return@mapNotNull null
            HarvestedSong(
                slug = song.slug,
                artist = song.artist,
                title = song.title,
                url = song.url,
                status = song.status,
                dataBlob = song.dataBlob ?: return@mapNotNull null,
                alphaGroup = AllSongsGrouping.alphabeticalGroup(song.title),
                modes = HarvestedSong.encodeModes(currentDb.songDao().getModesForSlug(song.slug)),
                harvestedAt = now
            )
        }
        userDb.withTransaction {
            dao.adopt(entries)
            dao.setFlag(HarvestLedgerMeta(LEGACY_ADOPTION_KEY, now.toString()))
        }
    }

    /**
     * Rebuilds every ledgered harvest the catalog is missing, and returns how
     * many rows it actually wrote.
     *
     * A slug the catalog already carries is left alone: that copy is newer and
     * carries a complexity rating a harvest never produces. Idempotent, so a
     * redundant pass after an interrupted install costs one query per entry.
     */
    suspend fun replay(catalogDb: AppDatabase, userDb: UserDataDatabase): Int {
        var restored = 0
        for (entry in userDb.harvestLedgerDao().getAll()) {
            if (catalogDb.songDao().songExists(entry.slug)) continue
            val song = entry.toSong()
            val browseModes = entry.modeSet().map { mode -> SongBrowseMode(entry.slug, mode) }
            catalogDb.withTransaction {
                catalogDb.songDao().insertSong(song)
                catalogDb.songDao().upsertBrowseEntry(
                    SongBrowseEntry(
                        slug = entry.slug,
                        artist = entry.artist,
                        title = entry.title,
                        alphaGroup = entry.alphaGroup,
                        complexityRating = null,
                        complexityBucket = null
                    )
                )
                catalogDb.songDao().deleteBrowseModes(entry.slug)
                if (browseModes.isNotEmpty()) {
                    catalogDb.songDao().insertBrowseModes(browseModes)
                }
            }
            restored++
        }
        return restored
    }

    /**
     * Which of [slugs] the staged catalog already carries.
     *
     * Asks about the candidates rather than reading forty thousand slugs back:
     * the question is only ever about a bounded handful. The staged file is not
     * the open database, so it is read directly, with the same read-only handle
     * DatabaseDownloader.ensureBrowseSchema already opens it with.
     */
    private fun slugsPresent(databaseFile: File, slugs: List<String>): Set<String> {
        if (!databaseFile.exists() || slugs.isEmpty()) return emptySet()
        val present = HashSet<String>()
        SQLiteDatabase.openDatabase(
            databaseFile.path,
            null,
            SQLiteDatabase.OPEN_READONLY
        ).use { staged ->
            // Chunked to stay well inside SQLite's bound-variable ceiling.
            for (chunk in slugs.chunked(200)) {
                val placeholders = chunk.joinToString(",") { "?" }
                staged.rawQuery(
                    "SELECT slug FROM songs WHERE slug IN ($placeholders)",
                    chunk.toTypedArray()
                ).use { cursor ->
                    while (cursor.moveToNext()) present.add(cursor.getString(0))
                }
            }
        }
        return present
    }
}

File: ios/Packages/AcquiringKit/Sources/AcquiringCatalog/HarvestLedger.swift
import AcquiringCore
import Foundation
import GRDB

/// One manually harvested song, held in full so it can be rebuilt into a fresh
/// catalog without contacting Hooktheory again.
public struct HarvestLedgerEntry: Sendable, Equatable {
    public let slug: String
    public let artist: String?
    public let title: String?
    public let url: String
    public let status: String
    public let payload: Data
    public let alphaGroup: String
    public let modes: Set<String>

    public init(
        slug: String,
        artist: String?,
        title: String?,
        url: String,
        status: String,
        payload: Data,
        alphaGroup: String,
        modes: Set<String>
    ) {
        self.slug = slug
        self.artist = artist
        self.title = title
        self.url = url
        self.status = status
        self.payload = payload
        self.alphaGroup = alphaGroup
        self.modes = modes
    }
}

/// User-owned record of every manually harvested song.
///
/// A catalog install replaces `catalog.db` as a whole file
/// (`CatalogCoordinator.replaceLiveDatabase(with:)`), so a harvested row written
/// only into the catalog is destroyed by the next download. This ledger lives in
/// its own file in its own directory — deliberately *outside*
/// `CatalogConfiguration.directoryURL`, which the install writes into and
/// `prepare()` sweeps — and is replayed into the catalog after each swap.
///
/// Unlike the catalog directory this one is not excluded from backup: a harvest
/// cannot be re-downloaded.
public final class HarvestLedger: Sendable {
    private let queue: DatabaseQueue

    public let directoryURL: URL

    public init(directoryURL: URL, fileManager: FileManager = .default) throws {
        self.directoryURL = directoryURL
        try fileManager.createDirectory(at: directoryURL, withIntermediateDirectories: true)
        var configuration = Configuration()
        configuration.label = "AcquiringHarvestLedger"
        queue = try DatabaseQueue(
            path: directoryURL.appending(path: "harvests.db").path,
            configuration: configuration
        )
        try queue.write { db in try db.execute(sql: Self.createSQL) }
    }

    /// Records a manual harvest. Re-harvesting the same slug replaces the entry,
    /// matching `CatalogCoordinator.writeHarvested(...)`.
    public func record(
        slug: String,
        artist: String?,
        title: String?,
        url: String,
        status: String,
        payload: Data,
        alphaGroup: String,
        modes: Set<String>
    ) throws {
        try queue.write { db in
            try Self.write(
                HarvestLedgerEntry(
                    slug: slug,
                    artist: artist,
                    title: title,
                    url: url,
                    status: status,
                    payload: payload,
                    alphaGroup: alphaGroup,
                    modes: modes
                ),
                replacingExisting: true,
                in: db
            )
        }
    }

    /// Takes ownership of songs harvested before the ledger existed. Never
    /// displaces an entry the user's own harvest already wrote.
    public func adopt(_ entries: [HarvestLedgerEntry]) throws {
        guard !entries.isEmpty else { return }
        try queue.write { db in
            for entry in entries {
                try Self.write(entry, replacingExisting: false, in: db)
            }
        }
    }

    public func entries() throws -> [HarvestLedgerEntry] {
        try queue.read { db in
            try Row.fetchAll(db, sql: """
                SELECT slug, artist, title, url, status, dataBlob, alphaGroup, modes
                FROM harvested_songs ORDER BY harvestedAt
                """).map(Self.entry(from:))
        }
    }

    public func isEmpty() throws -> Bool {
        try queue.read { db in try Int.fetchOne(db, sql: "SELECT COUNT(*) FROM harvested_songs") ?? 0 } == 0
    }

    /// The legacy sweep is a one-time amnesty for harvests made before this
    /// ledger shipped. Running it on every install would permanently resurrect
    /// songs the catalog later prunes on purpose. The flag lives beside the rows
    /// it guards so the two cannot drift apart.
    public func hasAdoptedLegacyHarvests() throws -> Bool {
        try queue.read { db in
            try String.fetchOne(db, sql: "SELECT value FROM harvest_ledger_meta WHERE key = ?", arguments: [Self.legacyAdoptionKey]) != nil
        }
    }

    public func markLegacyHarvestsAdopted() throws {
        try queue.write { db in
            try db.execute(
                sql: "INSERT OR REPLACE INTO harvest_ledger_meta (key, value) VALUES (?, ?)",
                arguments: [Self.legacyAdoptionKey, ISO8601DateFormatter().string(from: Date())]
            )
        }
    }

    private static let legacyAdoptionKey = "legacyHarvestsAdopted"

    private static func write(_ entry: HarvestLedgerEntry, replacingExisting: Bool, in db: Database) throws {
        try db.execute(
            sql: """
                INSERT OR \(replacingExisting ? "REPLACE" : "IGNORE") INTO harvested_songs
                    (slug, artist, title, url, status, dataBlob, alphaGroup, modes, harvestedAt)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
                """,
            arguments: [
                entry.slug, entry.artist, entry.title, entry.url, entry.status,
                entry.payload, entry.alphaGroup, encode(modes: entry.modes),
                Date().timeIntervalSince1970
            ]
        )
    }

    private static func entry(from row: Row) -> HarvestLedgerEntry {
        HarvestLedgerEntry(
            slug: row["slug"],
            artist: row["artist"],
            title: row["title"],
            url: row["url"],
            status: row["status"],
            payload: row["dataBlob"],
            alphaGroup: row["alphaGroup"],
            modes: decodeModes(row["modes"] ?? "")
        )
    }

    static func encode(modes: Set<String>) -> String {
        modes.sorted().joined(separator: "\n")
    }

    static func decodeModes(_ value: String) -> Set<String> {
        Set(value.split(separator: "\n").map(String.init))
    }

    private static let createSQL = """
        PRAGMA user_version = 1;
        CREATE TABLE IF NOT EXISTS harvested_songs (
            slug TEXT NOT NULL PRIMARY KEY, artist TEXT, title TEXT,
            url TEXT NOT NULL, status TEXT NOT NULL, dataBlob BLOB NOT NULL,
            alphaGroup TEXT NOT NULL, modes TEXT NOT NULL, harvestedAt REAL NOT NULL
        );
        CREATE TABLE IF NOT EXISTS harvest_ledger_meta (
            key TEXT NOT NULL PRIMARY KEY, value TEXT NOT NULL
        );
        """
}
